File: src/mitris/sim/core/simulation/Simulator.java
package mitris.sim.core.simulation;

import java.util.Collection;

import mitris.sim.core.modeling.InPort;
import mitris.sim.core.modeling.OutPort;
import mitris.sim.core.modeling.api.DevsAtomic;
import mitris.sim.core.simulation.api.SimulationClock;

/**
 *
 * @author José Luis Risco Martín
 */
public class Simulator extends AbstractSimulator {

    protected DevsAtomic model;
    
    public Simulator(SimulationClock clock, DevsAtomic model) {
        super(clock);
        this.model = model;
    }
    
    public void initialize() {
    	model.initialize();
        tL = clock.getTime();
        tN = tL + model.ta();    	
    }
    
   public double ta() {
        return model.ta();
    }

    public void deltfcn() {
        double t = clock.getTime();
        boolean isInputEmpty = model.isInputEmpty();
        if (isInputEmpty && t != tN) {
            return;
        } else if (!isInputEmpty && t == tN) {
            double e = t - tL;
            model.setSigma(model.getSigma() - e);
            model.deltcon(e);
        } else if (isInputEmpty && t == tN) {
            model.deltint();
        } else if (!isInputEmpty && t != tN) {
            double e = t - tL;
            model.setSigma(model.getSigma() - e);
            model.deltext(e);
        }
        tL = t;
        tN = tL + model.ta();
    }

    public void lambda() {
        if (clock.getTime() == tN) {
            model.lambda();
        }
    }

    public void clear() {
        Collection<InPort<?>> inPorts;
        inPorts = model.getInPorts();
        for (InPort<?> port : inPorts) {
            port.clear();
        }
        Collection<OutPort<?>> outPorts;
        outPorts = model.getOutPorts();
        for (OutPort<?> port : outPorts) {
            port.clear();
        }
    }
    
    public DevsAtomic getModel() {
    	return model;
    }
    
}

File: simulation/simulator.go
package simulation

import "github.com/pointlesssoft/godevs/modeling"

type Simulator struct {
	AbstractSimulator
	model modeling.AtomicInterface
}

func NewSimulator(clock *Clock, model modeling.AtomicInterface) *Simulator {
	s := Simulator{*NewAbstractSimulator(clock), model}
	return &s
}

func (s *Simulator) Initialize() {
	s.model.Initialize()
	s.SetTL((*s.GetClock()).GetTime())
	s.SetTN(s.GetTL() + s.model.TA())
}

func (s *Simulator) Exit() {
	s.model.Exit()
}

func (s *Simulator) TA() float64 {
	return s.model.TA()
}

func (s *Simulator) DeltFcn() {
	t := (*s.GetClock()).GetTime()
	isInputEmpty := s.model.IsInputEmpty()
	if !isInputEmpty || t == s.GetTN() {
		// CASE 1: atomic model timed out and no messages were received -> internal delta
		if isInputEmpty {
			s.model.DeltInt()
		} else {
			e := t - s.GetTL()
			// CASE 2: both atomic model timed out and messages were received -> confluent delta
			if t == s.GetTN() {
				s.model.DeltCon(e)
			// CASE 3: only messages were received -> external delta
			} else {
				s.model.DeltExt(e)
			}
		}
		s.SetTL(t)
		s.SetTN(t + s.model.TA())
	}
}

func (s *Simulator) Clear() {
	for _, ports := range [][]*modeling.Port{s.model.GetInPorts(), s.model.GetOutPorts()} {
		for _, port := range ports {
			port.Clear()
		}
	}
}

func (s *Simulator) Lambda() {
	if s.GetClock().GetTime() == s.GetTN() {
		s.model.Lambda()
	}
}

func (s *Simulator) GetModel() modeling.ComponentInterface {
	return s.model.GetComponent()
}
